include/otus/matrix.hpp

```diff
@@ -0,0 +1,221 @@
+    class Iterator;
+    /// TupleHash used as Hash function for std::unordered_map
+    class TupleHash;
+    /// Using these Layouts for access to other Layouts in the matrix
+    template <size_t N, typename... Types>
+    class Layout;
+    /// Using this Layout as Smart Object for get/set values of the matrix
+    template <typename... Types>
+    class Layout<0, Types...>;
+
+    using TupleKey = typename generate_tuple_type<size_t, Dimension>::type;
+    using Contanter = std::unordered_map<TupleKey, T, TupleHash>;
+    using NextLayout = Layout<Dimension - 1, size_t>;
+
+    Contanter elements_;
+    const T defaultValue_{DefaultValue};
+
+  public:
+    Matrix() = default;
+    ~Matrix() = default;
+    Matrix(const Matrix &other) noexcept : elements_(other.elements_) {}
+    Matrix(Matrix &&other) noexcept : elements_(std::move(other.elements_)) {}
+
+    Matrix(std::initializer_list<std::pair<const TupleKey, T>> list) {
+        elements_.reserve(list.size());
+        for (const auto &element : list) {
+            elements_.emplace(element);
+        }
+    }
+
+    Matrix &operator=(const Matrix &other) {
+        elements_ = other.elements_;
+        return *this;
+    }
+    Matrix &operator=(Matrix &&other) noexcept {
+        elements_ = std::move(other.elements_);
+        return *this;
+    }
+
+    bool operator==(const Matrix &other) const { return elements_ == other.elements_; }
+    bool operator!=(const Matrix &other) const { return !(*this == other); }
+
+    NextLayout operator[](size_t idx) { return NextLayout(idx, elements_); }
+    const NextLayout operator[](size_t idx) const { return NextLayout(idx, elements_); }
+
+    /// Return an input iterator to the beginning
+    /// @return Input iterator to the begining
+    auto begin() const noexcept { return Iterator(elements_.cbegin()); }
+
+    /// Return an input iterator to the end
+    /// @return Input iterator to the end
+    auto end() const noexcept { return Iterator(elements_.cend()); }
+
+    /// Get real count of elements in matrix
+    /// @return count of elements
+    size_t size() const noexcept { return elements_.size(); }
+
+    /// Clears the mapped matrix.
+    void clear() noexcept { elements_.clear(); }
+};
+
+// ****************************
+// * Struct Matrix::TupleHash *
+// ****************************
+// Boost for combining hash values
+// https://www.boost.org/doc/libs/1_38_0/doc/html/hash/reference.html#boost.hash_combine
+template <typename T, T DefaultValue, size_t Dimension>
+class Matrix<T, DefaultValue, Dimension>::TupleHash {
+    inline size_t get_hash(size_t &seed, size_t value) const {
+        return std::hash<size_t>()(value) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+    }
+
+    template <std::size_t... Indices>
+    void combine_hash(size_t &seed, const TupleKey &tuple, std::index_sequence<Indices...>) const {
+        using swallow = size_t[];
+        (void)swallow{(seed ^= get_hash(seed, std::get<Indices>(tuple)))...};
```

Comment:
```suggestion
        ((seed ^= get_hash(seed, std::get<Indices>(tuple))), ...);
```
Вы ведь это хотели сделать тут? (правда, требует 17 стандарт). Я долго вглядывался в код, прежде чем понял, что тут происходит и зачем алиас на тип массива, приведенный к `void` :smiley: